File: livraria-frontend/src/services/livroService.ts
import { api } from "../api/axios";
import type{ Livro } from "../types/Livro";
import type { ResponseModel } from "../types/ResponseModel";

// GET - Listar livros
export async function listarLivros() {
  const response = await api.get<ResponseModel<Livro[]>>(
    "/Livro/ListarLivros"
  );
  return response.data;
}

// GET - Buscar livro por ID
export async function buscarLivroPorId(idLivro: number) {
  const response = await api.get<ResponseModel<Livro>>(
    `/Livro/BuscarLivroPorId/${idLivro}`
  );
  return response.data;
}

// GET - Buscar livro por ID do autor
export async function buscarLivroPorIdAutor(idAutor: number) {
  const response = await api.get<ResponseModel<Livro>>(
    `/Livro/BuscarLivroPorIdAutor/${idAutor}`
  );
  return response.data;
}

// POST - Criar livro
export async function criarLivro(livro: {
  titulo: string;
  isbn: string;
  anoPublicacao: number;
  autorId: number;
}) {
  const response = await api.post<ResponseModel<Livro>>(
    "/Livro/CriarLivro",
    livro
  );
  return response.data;
}

// PUT - Editar livro
export async function editarLivro(livro: {
  idLivro: number;
  titulo: string;
  isbn: string;
  anoPublicacao: number;
  autorId: number;
}) {
  const response = await api.put<ResponseModel<Livro>>(
    "/Livro/EditarLivro",
    livro
  );
  return response.data;
}

// DELETE - Excluir livro
export async function excluirLivro(idLivro: number) {
  const response = await api.delete<ResponseModel<Livro>>(
    "/Livro/ExcluirLivro",
    { params: { idLivro } }
  );
  return response.data;
}

File: livraria-backend/Services/Livro/LivroService.cs
using System.Collections.Generic;
using Livraria.Data;
using Livraria.Dto.Autor;
using Livraria.Dto.Livro;
using Livraria.Models;
using Microsoft.EntityFrameworkCore;

namespace Livraria.Services.Livro
{
    public class LivroService : ILivroInterface
    {
        private readonly AppDbContext _context;
        public LivroService(AppDbContext context)
        {
            _context = context;
        }


        public async Task<ResponseModel<LivroModel>> BuscarLivroPorId(int idLivro)
        {
            ResponseModel<LivroModel> resposta = new ResponseModel<LivroModel>();
            try
            {
                var livro = await _context.Livros.FirstOrDefaultAsync(livroBanco => livroBanco.Id == idLivro);
                if(livro == null)
                {
                    resposta.Mensagem = "Nenhum livro encontrado!";
                    return resposta;
                }
                resposta.Dados = livro;
                resposta.Mensagem = "Livro encontrado!";
                return resposta;


            }
            catch (Exception ex)
            {
                resposta.Mensagem = ex.Message;
                resposta.Status = false;
                return resposta;

            }

        }

        public async Task <ResponseModel<List<LivroModel>>> BuscarLivroPorIdAutor(int idAutor)
        {
			ResponseModel<List<LivroModel>> resposta = new ResponseModel<List<LivroModel>>();
			try
			{
				var livro = await _context.Livros.Include(a => a.Autor).Where(livroBanco => livroBanco.Autor.Id == idAutor).ToListAsync();

				if (livro == null)
				{
					resposta.Mensagem = "Nenhum registro encontrado";
					return resposta;
				}
                resposta.Dados = livro;
				resposta.Mensagem = "Autor Localizado!";
				return resposta;



			}
			catch (Exception ex)
			{
				resposta.Mensagem = ex.Message;
				resposta.Status = false;
				return resposta;
			}
		}

        public async Task<ResponseModel<List<LivroModel>>> CriarLivro(LivroDto livroDto)
        {
			ResponseModel<List<LivroModel>> resposta = new ResponseModel<List<LivroModel>>();

			try
			{
                var autor = await _context.Autor.FirstOrDefaultAsync(autorBanco => autorBanco.Id == livroDto.Autor.Id);
                if (autor == null)
                {
                    resposta.Mensagem = "Nenhum registro de autor localizado!";
                    return resposta;
                }
                var livro = new LivroModel()
                {
                    Titulo = livroDto.Titulo,
                    Autor = autor,
                };
                _context.Add(livro);
                await _context.SaveChangesAsync();
                resposta.Dados = await _context.Livros.Include(a => a.Autor).ToListAsync();
                return resposta;
			}
			catch (Exception ex)
			{
				resposta.Mensagem = ex.Message;
				resposta.Status = false;
				return resposta;

			}
		}

        public async Task<ResponseModel<List<LivroModel>>> EditarLivro(LivroEditarDto livroEditarDto)
        {
			ResponseModel<List<LivroModel>> resposta = new ResponseModel<List<LivroModel>>();
			try
			{
                var livro = await _context.Livros.Include(a => a.Autor).FirstOrDefaultAsync(livroBanco => livroBanco.Id == livroEditarDto.Id);

                var autor = await _context.Autor.FirstOrDefaultAsync(autorBanco => autorBanco.Id == livroEditarDto.Autor.Id);

				if (autor == null)
				{
					resposta.Mensagem = "Nenhum registro de autor localizado!";
					return resposta;
				}
				if (livro == null)
				{
					resposta.Mensagem = "Nenhum registro de livro localizado!";
					return resposta;
				}
				livro.Titulo = livroEditarDto.Titulo;
				livro.Autor = autor;

				_context.Update(livro);
				await _context.SaveChangesAsync();

				resposta.Dados = await _context.Livros.ToListAsync();
				return resposta;


			}
			catch (Exception ex)
			{
				resposta.Mensagem = ex.Message;
				resposta.Status = false;
				return resposta;

			}
		}

        public async Task<ResponseModel<List<LivroModel>>> ExcluirLivro(int idLivro)
        {
			ResponseModel<List<LivroModel>> resposta = new ResponseModel<List<LivroModel>>();

			try
			{
				var livro = await _context.Livros.FirstOrDefaultAsync(livroBanco => livroBanco.Id == idLivro);

				if (livro == null)
				{
					resposta.Mensagem = "Nenhum livro localizado!";
					return resposta;
				}
				_context.Remove(livro);
				await _context.SaveChangesAsync();

				resposta.Dados = await _context.Livros.ToListAsync();
				resposta.Mensagem = "Livro removido com sucesso!";
				return resposta;

			}
			catch (Exception ex)
			{
				resposta.Mensagem = ex.Message;
				resposta.Status = false;
				return resposta;

			}
		}

        public async Task<ResponseModel<List<LivroModel>>> ListarLivros()
        {
			ResponseModel<List<LivroModel>> resposta = new ResponseModel<List<LivroModel>>();
			try
			{
                var livros = await _context.Livros.ToListAsync();

                resposta.Dados = livros;
                resposta.Mensagem = "Aqui estão todos os livros!";
                return resposta;
				


			}
			catch (Exception ex)
			{
				resposta.Mensagem = ex.Message;
				resposta.Status = false;
				return resposta;

			}
		}
    }
}
